be/src/formats/parquet/meta_helper.cpp

```diff
@@ -120,32 +120,57 @@ void IcebergMetaHelper::_init_field_mapping() {
     }
 }
 
-bool IcebergMetaHelper::_is_valid_type(const ParquetField* parquet_field,
-                                       const TIcebergSchemaField* field_schema) const {
+bool IcebergMetaHelper::_is_valid_type(const ParquetField* parquet_field, const TIcebergSchemaField* field_schema,
+                                       const TypeDescriptor* type_descriptor) const {
     // only check for complex type now
     // if complex type has none valid subfield, we will treat this struct type as invalid type.
     if (!parquet_field->type.is_complex_type()) {
         return true;
     }
 
+    if (parquet_field->type.type != type_descriptor->type) {
+        // complex type mismatched
+        return false;
+    }
+
     bool has_valid_child = false;
 
-    std::unordered_map<int32_t, const TIcebergSchemaField*> field_id_2_iceberg_schema{};
-    for (const auto& field : field_schema->children) {
-        field_id_2_iceberg_schema.emplace(field.field_id, &field);
-    }
+    if (parquet_field->type.is_array_type() || parquet_field->type.is_map_type()) {
+        for (size_t idx = 0; idx < parquet_field->children.size(); idx++) {
+            if (_is_valid_type(&parquet_field->children[idx], &field_schema->children[idx],
+                               &type_descriptor->children[idx])) {
+                has_valid_child = true;
+                break;
+            }
+        }
+    } else if (parquet_field->type.is_struct_type()) {
+        std::unordered_map<int32_t, const TIcebergSchemaField*> field_id_2_iceberg_schema{};
+        for (const auto& field : field_schema->children) {
+            field_id_2_iceberg_schema.emplace(field.field_id, &field);
+        }
 
-    // start to check struct type
-    for (const auto& child_parquet_field : parquet_field->children) {
-        auto it = field_id_2_iceberg_schema.find(child_parquet_field.field_id);
-        if (it == field_id_2_iceberg_schema.end()) {
-            continue;
+        std::unordered_map<int32_t, const TypeDescriptor*> field_id_2_type{};
+        for (size_t idx = 0; idx < type_descriptor->children.size(); idx++) {
+            field_id_2_type.emplace(parquet_field->children[idx].field_id, &type_descriptor->children[idx]);
```

Comment:
oh, i think this code is not correct, because in the test case, col->d is not exist in parquet field, so the field_id_2_type should be empty, but for this code, field_id_2_type will never be empty.

Reply:
Ok i added a fix for that, but this still doesn't solve that the subfield is not being set to null, this just skips the whole column and sets it to null which is incorrect:

```
[ RUN      ] IcebergSchemaEvolutionTest.TestStructEvolutionPadNull
../test/formats/parquet/iceberg_schema_evolution_file_reader_test.cpp:267: Failure
Expected equality of these values:
  "[{d:NULL}]"
  chunk->debug_row(0)
    Which is: "[NULL]"
[  FAILED  ] IcebergSchemaEvolutionTest.TestStructEvolutionPadNull (0 ms)
[----------] 1 test from IcebergSchemaEvolutionTest (0 ms total)
```

Reply:
I got what you mean, let me check with Trino's behavior.

Reply:
I think currently implementation is correct.
```sql
mysql> select * from struct_add_subfield;
+------+---------------------------+
| id   | col_struct                |
+------+---------------------------+
|    1 | {"age":15,"name":null}    |
|    2 | {"age":15,"name":"smith"} |
+------+---------------------------+
2 rows in set (0.07 sec)
```

I've built a table, and the subfield `col_struct.name` is newly added, which means `id=1`'s parquet didn't contain `col_struct.name`.

In spark:
```sql
spark-sql (chendingchao_ice)> select col_struct.name from struct_add_subfield;
NULL
smith
```

In trino:
```sql
trino:chendingchao_ice> select col_struct.name from struct_add_subfield;
 name
-------
 NULL
 smith
(2 rows)
```

In sr, your currently implementation:
```sql
mysql> select col_struct.name from struct_add_subfield;
+-------+
| name  |
+-------+
| NULL  |
| smith |
+-------+
2 rows in set (0.06 sec)
```

Reply:
Ok i just checked with Athena and it seems you are right it does this behavior as well, let me update the test.

Reply:
Good job, and thank you!